Building with mingw warns 'examples/io/vsgcluster/Broadcaster.h:32: warning: "NOMINMAX" redefined'

Cross compiling vsgExamples with gcc for Windows fails with the following error:
```
[  121s] cd /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster && /usr/bin/x86_64-w64-mingw32-g++  -DVSGXCHANGE_SHARED_LIBRARY -DVSG_SHARED_LIBRARY -DvsgXchange_FOUND  -O2 -g -pipe -Wall -fexceptions --param=ssp-buffer-size=4 -mms-bitfields -O2 -g -DNDEBUG   -Wall -Wparentheses -Wno-long-long -Wno-import -Wreturn-type -Wmissing-braces -Wunknown-pragmas -Wmaybe-uninitialized -Wshadow -Wunused -Wno-misleading-indentation -Wextra -std=gnu++17 -o CMakeFiles/vsgcluster.dir/Packet.cpp.obj -c /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp
[  121s] In file included from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.h:7,
[  121s]                  from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp:5:
[  121s] /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Broadcaster.h:32: warning: "NOMINMAX" redefined
[  121s]    32 | #    define NOMINMAX
[  121s]       | 
[  121s] In file included from /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/x86_64-w64-mingw32/bits/c++config.h:522,
[  121s]                  from /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/iostream:38,
[  121s]                  from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp:2:
[  121s] /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45: note: this is the location of the previous definition
[  121s]    45 | #define NOMINMAX 1
[  121s]       | 
[  121s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939'
[  128s] make[2]: Entering directory '/home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939'
```

It looks like in `io/vsgcluster/Broadcaster.h:32` the macro NOMINMAX is defined unconditionally, but already defined by the mingw headers. If this value is needed as a fallback for other compilers that do not define this macro, it should be defined with `#ifndef NOMINMAX ... #endif`.

Could you add the #idef to quite the warning and generate a PR for it.
Thanks.

On Mon, 5 Sept 2022 at 08:48, Ralf Habacker ***@***.***>
wrote:

> Cross compiling vsgExamples with gcc for Windows fails with the following
> error:
>
> [  121s] cd /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster && /usr/bin/x86_64-w64-mingw32-g++  -DVSGXCHANGE_SHARED_LIBRARY -DVSG_SHARED_LIBRARY -DvsgXchange_FOUND  -O2 -g -pipe -Wall -fexceptions --param=ssp-buffer-size=4 -mms-bitfields -O2 -g -DNDEBUG   -Wall -Wparentheses -Wno-long-long -Wno-import -Wreturn-type -Wmissing-braces -Wunknown-pragmas -Wmaybe-uninitialized -Wshadow -Wunused -Wno-misleading-indentation -Wextra -std=gnu++17 -o CMakeFiles/vsgcluster.dir/Packet.cpp.obj -c /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp
> [  121s] In file included from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.h:7,
> [  121s]                  from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp:5:
> [  121s] /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Broadcaster.h:32: warning: "NOMINMAX" redefined
> [  121s]    32 | #    define NOMINMAX
> [  121s]       |
> [  121s] In file included from /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/x86_64-w64-mingw32/bits/c++config.h:522,
> [  121s]                  from /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/iostream:38,
> [  121s]                  from /home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939/examples/io/vsgcluster/Packet.cpp:2:
> [  121s] /usr/lib64/gcc/x86_64-w64-mingw32/10.3.0/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45: note: this is the location of the previous definition
> [  121s]    45 | #define NOMINMAX 1
> [  121s]       |
> [  121s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939'
> [  128s] make[2]: Entering directory '/home/abuild/rpmbuild/BUILD/vsgExamples-0.0.0+20220713+git.3ac4939'
>
> It looks like in io/vsgcluster/Broadcaster.h:32 the macro NOMINMAX is
> defined unconditionally, but already defined by the mingw headers. If this
> value is needed as a fallback for other compilers that do not define this
> macro, it should be defined with #ifndef NOMINMAX ... #endif.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vsg-dev/vsgExamples/issues/170>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEGUAUBVQ37766QNMMOSTV4WQT3ANCNFSM6AAAAAAQEXGWPI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/examples/io/vsgcluster/Broadcaster.cpp
+++ b/examples/io/vsgcluster/Broadcaster.cpp
@@ -16,6 +16,8 @@
 *  THE SOFTWARE.
 */
 
+#include "Broadcaster.h"
+
 #include <fcntl.h>
 #include <iostream>
 #include <stdio.h>
@@ -55,16 +57,11 @@
 #    include <sys/sockio.h>
 #    include <unistd.h>
 #elif defined(WIN32)
-#    define NOMINMAX
-#    include <stdio.h>
-#    include <winsock.h>
 #elif defined(__hpux)
 #    include <unistd.h>
 #else
 #    error Teach me how to build on this system
 #endif
-
-#include "Broadcaster.h"
 
 #define _VERBOSE 1
 

--- a/examples/io/vsgcluster/Broadcaster.h
+++ b/examples/io/vsgcluster/Broadcaster.h
@@ -28,7 +28,7 @@
 // Class definition for broadcasting a buffer to a LAN
 //
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
 #    define NOMINMAX
 #    include <winsock.h>
 #else
@@ -57,14 +57,14 @@ private:
 
     std::string _ifr_name;
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     SOCKET _so;
 #else
     int _so;
 #endif
     bool _initialized;
     short _port;
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     SOCKADDR_IN saddr;
 #else
     struct sockaddr_in saddr;

--- a/examples/io/vsgcluster/Broadcaster.h
+++ b/examples/io/vsgcluster/Broadcaster.h
@@ -29,7 +29,9 @@
 //
 
 #if defined(WIN32) && !defined(__CYGWIN__)
+#ifndef NOMINMAX
 #    define NOMINMAX
+#endif
 #    include <winsock.h>
 #else
 #    include <netinet/in.h>

--- a/examples/io/vsgcluster/Receiver.cpp
+++ b/examples/io/vsgcluster/Receiver.cpp
@@ -19,7 +19,7 @@
 #include <fcntl.h>
 #include <stdio.h>
 #include <sys/types.h>
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
 #    define NOMINMAX
 #    include <winsock.h>
 #else
@@ -42,7 +42,7 @@ Receiver::Receiver(uint16_t port) :
     _initialized(false),
     _port(port)
 {
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     WORD version = MAKEWORD(1, 1);
     WSADATA wsaData;
     // First, we start up Winsock
@@ -52,7 +52,7 @@ Receiver::Receiver(uint16_t port) :
 
 Receiver::~Receiver(void)
 {
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     closesocket(_so);
 
     WSACleanup();
@@ -75,7 +75,7 @@ bool Receiver::init(void)
         return false;
     }
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     const BOOL on = TRUE;
     setsockopt(_so, SOL_SOCKET, SO_REUSEADDR, (const char*)&on, sizeof(int));
 #else
@@ -85,14 +85,14 @@ bool Receiver::init(void)
 
     saddr.sin_family = AF_INET;
     saddr.sin_port = htons(_port);
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     saddr.sin_addr.s_addr = htonl(INADDR_ANY);
 #else
     saddr.sin_addr.s_addr = 0;
 #endif
 
     // set up a 1 second timeout.
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     DWORD tv = 1000; // 1 sec in ms
     if (setsockopt(_so, SOL_SOCKET, SO_RCVTIMEO, reinterpret_cast<const char*>(&tv), sizeof(DWORD)))
     {
@@ -141,7 +141,7 @@ unsigned int Receiver::receive(void* buffer, const unsigned int buffer_size)
     FD_ZERO(&fdset);
     FD_SET(_so, &fdset);
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
 
     int read_bytes = recvfrom(_so, (char*)buffer, buffer_size, 0, (sockaddr*)&saddr, &size);
 

--- a/examples/io/vsgcluster/Receiver.cpp
+++ b/examples/io/vsgcluster/Receiver.cpp
@@ -16,13 +16,12 @@
 *  THE SOFTWARE.
 */
 
+#include "Receiver.h"
+
 #include <fcntl.h>
 #include <stdio.h>
 #include <sys/types.h>
-#if defined(WIN32) && !defined(__CYGWIN__)
-#    define NOMINMAX
-#    include <winsock.h>
-#else
+#if !defined(WIN32)
 #    include <arpa/inet.h>
 #    include <errno.h>
 #    include <netdb.h>
@@ -33,8 +32,6 @@
 #    include <unistd.h>
 #endif
 #include <string.h>
-
-#include "Receiver.h"
 
 #include <iostream>
 

--- a/examples/io/vsgcluster/Receiver.h
+++ b/examples/io/vsgcluster/Receiver.h
@@ -25,7 +25,7 @@
 // Class definition for the recipient of a broadcasted message
 //
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
 #    define NOMINMAX
 #    include <winsock.h>
 #else
@@ -48,7 +48,7 @@ private:
 private:
     virtual ~Receiver();
 
-#if defined(_WIN32) && !defined(__CYGWIN__)
+#if defined(WIN32) && !defined(__CYGWIN__)
     SOCKET _so;
     SOCKADDR_IN saddr;
 #else

--- a/examples/io/vsgcluster/Receiver.h
+++ b/examples/io/vsgcluster/Receiver.h
@@ -26,7 +26,9 @@
 //
 
 #if defined(WIN32) && !defined(__CYGWIN__)
+#ifndef NOMINMAX
 #    define NOMINMAX
+#endif
 #    include <winsock.h>
 #else
 #    include <netinet/in.h>

